File: backend/app/schemas/pipeline.py
from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


class PipelineCreate(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    description: str | None = None
    dag_json: dict[str, Any] = Field(default_factory=lambda: {"nodes": [], "edges": []})


class PipelineUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None
    dag_json: dict[str, Any] | None = None
    is_active: bool | None = None


class PipelineResponse(BaseModel):
    id: UUID
    name: str
    description: str | None = None
    dag_json: dict[str, Any]
    schedule_cron: str | None = None
    is_active: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class RunTriggerResponse(BaseModel):
    run_id: UUID
    status: str


class PipelineRunResponse(BaseModel):
    id: UUID
    pipeline_id: UUID
    status: str
    triggered_by: str
    celery_task_id: str | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    rows_read: int | None = None
    rows_written: int | None = None
    target_dataset_id: UUID | None = None
    iceberg_snapshot_id: int | None = None
    error_message: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class RunLogsResponse(BaseModel):
    lines: list[str]
    next_since: int

File: frontend/src/types/pipeline.ts
export interface PipelineNode {
  id: string;
  type: "source" | "transform" | "output";
  dataset_id?: string;
  transform_type?: string;
  config?: Record<string, unknown>;
}

export interface PipelineEdge {
  from: string;
  to: string;
}

export interface PipelineDag {
  nodes: PipelineNode[];
  edges: PipelineEdge[];
}

export interface Pipeline {
  id: string;
  name: string;
  description: string | null;
  dag_json: PipelineDag;
  schedule_cron: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface PipelineRun {
  id: string;
  pipeline_id: string;
  status: "queued" | "running" | "success" | "failed" | "cancelled";
  triggered_by: string;
  celery_task_id: string | null;
  started_at: string | null;
  completed_at: string | null;
  rows_read: number | null;
  rows_written: number | null;
  target_dataset_id: string | null;
  iceberg_snapshot_id: number | null;
  error_message: string | null;
  created_at: string;
}

export interface RunLogs {
  lines: string[];
  next_since: number;
}
